File: packages/db/db_writer.py
"""
extractors/db_writer.py

Shared Postgres write helpers every Python extractor imports. This is the
concrete thing that makes the extractor contract's idempotency and
validation requirements load-bearing instead of aspirational — an
extractor author doesn't hand-write ingest/write SQL and hope it matches
the contract, they call these functions.

Owns exactly what the two shared tables need:
  - ingest()      -> a transaction that covers the ingested_files row AND the
                     forensic_records rows for one file, or neither
  - write_record() / write_records() -> validated insert into forensic_records

Deliberately does NOT own:
  - source-format parsing (each extractor's own code)
  - the `fields` sub-shape (each extractor owns and documents its own, per
    EXTRACTOR_CONTRACT.md #4)

Why `ingest()` is a context manager and not the old `ingest_file()`
-------------------------------------------------------------------
The previous API was two independent calls, each of which committed:

    file_hash, already = ingest_file(conn, ...)   # committed here
    if already: return
    records = parse(...)                          # <-- anything failing here
    write_records(conn, run_id, file_hash, records)   # committed here

and dedup was keyed on the ledger row merely existing:

    SELECT 1 FROM ingested_files WHERE file_hash = %s
    if found: skip this file entirely

Those two facts combine into permanent silent data loss. If anything went
wrong between the two commits, the ledger claimed the file was ingested while
none of its records existed, and every later run skipped it — reporting
success — forever. In a chain-of-custody tool, evidence disappeared and
nothing said so.

That was not a rare interruption case. All three extractors hit it on their
ordinary error paths: crash/main.py `continue`s when parse_ips_file fails,
mvt_iocs `return`s when alerts.json won't parse, ileapp_bridge `return`s when
every record in an artifact fails to normalize — each after ingest_file() had
already committed the ledger row.

The fix is structural rather than a rule to follow. `ingest()` owns the
transaction: the ledger row, the raw payload, the records, and the completion
flag all commit together or roll back together. There is no exposed call that
commits a ledger row on its own, so a caller cannot reconstruct the old
sequence by accident. `write_record`/`write_records` no longer commit at all
(they are also used inside a unit), which is what the caller-owns-the-boundary
change asks for and incidentally stops `write_record` from paying a commit per
row.

A future TypeScript extractor uses packages-ts/db-writer/dbWriter.ts, which
mirrors this file's semantics with a callback in place of the context
manager. It used to live under packages-ts/orchestrator/src/ alongside a
now-deleted, never-imported IngestionOrchestrator class; moving it kept the
one real thing in that directory and dropped the dead one, rather than
implying the orchestrator itself depends on it (it never did — see
main-orchestrator/main.ts, which writes nothing to Postgres directly and
only spawns the Python extractors that do).
"""

from __future__ import annotations

import hashlib
import sys
from contextlib import contextmanager
from pathlib import Path
from typing import Any, Iterator

import psycopg2
import psycopg2.extras

# Resolve packages-py directory
_EXTRACTORS_DIR = Path(__file__).resolve().parent
_PACKAGES_PY = _EXTRACTORS_DIR.parent
sys.path.insert(0, str(_PACKAGES_PY / "contracts"))

from normalized_record import NormalizedRecord  # noqa: E402


def compute_file_hash(path: str | Path) -> str:
    """sha256 of file contents — the idempotency key for ingested_files.
    Streamed in chunks so this doesn't load a large SMS attachment or
    gcloud log export fully into memory just to hash it."""
    h = hashlib.sha256()
    with open(path, "rb") as f:
        for chunk in iter(lambda: f.read(1024 * 1024), b""):
            h.update(chunk)
    return h.hexdigest()


class IngestUnit:
    """One file's worth of work, inside one transaction.

    Handed to the caller by `ingest()`. The caller checks
    `already_ingested`, then writes records and (optionally) the raw payload.
    Nothing here commits; `ingest()` commits once on clean exit.
    """

    def __init__(self, conn, run_id: str, file_path: Path, file_hash: str, already_ingested: bool):
        self._conn = conn
        self._run_id = run_id
        self.file_path = file_path
        self.file_hash = file_hash
        self.already_ingested = already_ingested
        self.records_written = 0

    def write(self, records: list[NormalizedRecord]) -> int:
        """Write validated records for this file. Callable more than once;
        counts accumulate into the ledger's record_count."""
        self._guard()
        written = write_records(self._conn, self._run_id, self.file_hash, records)
        self.records_written += written
        return written

    def write_one(self, record: NormalizedRecord) -> None:
        """Single-record form, for extractors that produce one record per file
        (crash reports). No longer more expensive than batching per commit,
        since neither commits."""
        self._guard()
        write_record(self._conn, self._run_id, self.file_hash, record)
        self.records_written += 1

    def set_raw_payload(self, payload: dict[str, Any]) -> None:
        """Attach the parsed payload to the ledger row.

        Extractors that can only build the payload after parsing (crash,
        mvt_iocs) used to insert `{}`, commit, parse, then UPDATE and commit
        again — a second window in which the ledger was wrong. Inside a unit
        this is just part of the same transaction.
        """
        self._guard()
        with self._conn.cursor() as cur:
            cur.execute(
                "UPDATE ingested_files SET raw_payload = %s WHERE file_hash = %s",
                (psycopg2.extras.Json(payload), self.file_hash),
            )

    def _guard(self) -> None:
        if self.already_ingested:
            raise RuntimeError(
                f"{self.file_path.name}: this file is already fully ingested "
                f"(file_hash {self.file_hash[:12]}). Check `already_ingested` and return "
                "before writing; writing here would duplicate records that are already "
                "committed, because the dedup key is the file hash and this file has one."
            )


@contextmanager
def ingest(
    conn,
    run_id: str,
    file_path: str | Path,
    source_type: str,
    raw_payload: dict[str, Any] | None = None,
) -> Iterator[IngestUnit]:
    """
    Atomic ingest of one file: ledger row + records + completion flag, or nothing.

    Usage::

        with ingest(conn, run_id, path, source_type=..., raw_payload=summary) as unit:
            if unit.already_ingested:
                pass            # dedup — prior run finished this file
            else:
                unit.write(records)

    Guarantees:

    - On clean exit the ledger row is marked complete with its record count and
      everything commits together.
    - On any exception the whole unit rolls back, including the ledger row, so
      the file has no trace in the ledger and the NEXT run retries it. That is
      the property the old two-commit API lacked.
    - `already_ingested` is True only when a previous run marked the file
      complete — never merely because a row exists.
    - An abandoned unit from an earlier crash (row present, not complete) is
      reclaimed: its orphaned records are deleted and it is re-ingested under
      the current run_id. Without this, rows stranded by the old code, or by a
      hard kill of the new code, would be retried but their partial records
      double-counted.

    Concurrency: the upsert below locks the ledger row for the duration of the
    transaction, so two extractors racing on the same file serialize instead of
    both deciding to write. The previous SELECT-then-INSERT was a read outside
    any lock; its `ON CONFLICT DO NOTHING` avoided a crash on the race but left
    both callers believing they should write records.
    """
    file_path = Path(file_path)
    file_hash = compute_file_hash(file_path)

    try:
        with conn.cursor() as cur:
            # Fast path for the common case on a re-run: the file is already
            # complete, so there is nothing to lock and nothing to write.
            #
            # This read is deliberately unlocked, which is safe because
            # ingest_complete is monotonic -- it goes FALSE -> TRUE exactly
            # once and never back, so a TRUE observed here cannot be
            # invalidated by a concurrent transaction. A FALSE might be stale,
            # which is why FALSE falls through to the locking upsert below
            # rather than being acted on.
            #
            # Without this, re-running against a large backup would issue a
            # no-op UPDATE per already-finished file, producing a dead tuple
            # and a row lock per file for no reason.
            cur.execute(
                "SELECT ingest_complete FROM ingested_files WHERE file_hash = %s",
                (file_hash,),
            )
            row = cur.fetchone()
            already_ingested = bool(row is not None and row[0])

        if already_ingested:
            # Nothing was modified, so there is nothing to

File: packages/db/dbWriter.js
"use strict";
/**
 * Postgres write helpers for TypeScript callers — the mirror of
 * packages/db/db_writer.py, and deliberately semantically identical
 * to it rather than merely similar.
 *
 * Why `ingest()` takes a callback instead of exposing `ingestFile()`
 * -----------------------------------------------------------------
 * This file previously offered `ingestFile()` / `writeRecord()` / `writeRecords()`
 * as independent calls, exactly like the Python module did, and inherited the
 * same defect: the `ingested_files` ledger row was written before the file's
 * `forensic_records`, dedup was keyed on that row merely existing
 *
 *     SELECT 1 FROM ingested_files WHERE file_hash = $1 LIMIT 1
 *     -> alreadyIngested: true, caller skips the file entirely
 *
 * and nothing tied the two writes into one transaction. Any failure in between
 * left a ledger row claiming a file was ingested while none of its records
 * existed. Every later run then skipped it, reported success, and the evidence
 * was permanently absent from a chain-of-custody database with no error
 * anywhere.
 *
 * The TypeScript version had it worse in one respect: it issued no BEGIN at
 * all, so under node-postgres's default autocommit every statement committed
 * on its own. There was not even a transaction to lose.
 *
 * `ingest()` now owns the boundary. It BEGINs, hands the caller a unit, and
 * COMMITs only after marking the ledger row complete with its record count —
 * or ROLLBACKs everything, ledger row included, so the next run retries the
 * file. No exported function commits a ledger row on its own, so the old
 * sequence cannot be reassembled by a caller.
 *
 * `writeRecord`/`writeRecords` are still exported for use inside a unit and no
 * longer manage transactions themselves.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAX_RECORDS_PER_STATEMENT = exports.IngestUnit = exports.DEFAULT_DB_URL = void 0;
exports.computeFileHash = computeFileHash;
exports.ingest = ingest;
exports.writeRecord = writeRecord;
exports.writeRecords = writeRecords;
exports.incompleteIngests = incompleteIngests;
const crypto = __importStar(require("node:crypto"));
const node_fs_1 = require("node:fs");
const contracts_1 = require("@verichron/contracts");
exports.DEFAULT_DB_URL = 'postgresql://forensics:forensics_dev_only@localhost:5432/forensics';
/**
 * sha256 of file contents — the idempotency key for ingested_files.
 *
 * Streamed rather than `readFile`d: this hashes SMS attachments and gcloud log
 * exports, and the previous implementation buffered the entire file into memory
 * just to hash it, which the Python side had already avoided. Node also caps a
 * single Buffer well below the size of a full backup artifact, so the old
 * version would throw outright on a large enough file.
 */
async function computeFileHash(filePath) {
    const hash = crypto.createHash('sha256');
    const stream = (0, node_fs_1.createReadStream)(filePath, { highWaterMark: 1024 * 1024 });
    return new Promise((resolve, reject) => {
        stream.on('data', (chunk) => hash.update(chunk));
        stream.on('error', reject);
        stream.on('end', () => resolve(hash.digest('hex')));
    });
}
/**
 * One file's worth of work, inside one transaction. Nothing here commits.
 */
class IngestUnit {
    client;
    runId;
    filePath;
    fileHash;
    alreadyIngested;
    recordsWritten = 0;
    constructor(client, runId, filePath, fileHash, 
    /** True only when a previous run marked this file COMPLETE. */
    alreadyIngested) {
        this.client = client;
        this.runId = runId;
        this.filePath = filePath;
        this.fileHash = fileHash;
        this.alreadyIngested = alreadyIngested;
    }
    /** Write validated records for this file. Callable more than once. */
    async write(records) {
        this.guard();
        const written = await writeRecords(this.client, this.runId, this.fileHash, records);
        this.recordsWritten += written;
        return written;
    }
    async writeOne(record) {
        this.guard();
        await writeRecord(this.client, this.runId, this.fileHash, record);
        this.recordsWritten += 1;
    }
    /**
     * Attach the parsed payload to the ledger row, in this same transaction.
     * For extractors that can only build the payload after parsing.
     */
    async setRawPayload(payload) {
        this.guard();
        await this.client.query('UPDATE ingested_files SET raw_payload = $1 WHERE file_hash = $2', [
            payload,
            this.fileHash,
        ]);
    }
    guard() {
        if (this.alreadyIngested) {
            throw new Error(`${this.filePath}: this file is already fully ingested (file_hash ` +
                `${this.fileHash.slice(0, 12)}). Check \`alreadyIngested\` and return before ` +
                `writing; writing here would duplicate records that are already committed, ` +
                `because the dedup key is the file hash and this file has one.`);
        }
    }
}
exports.IngestUnit = IngestUnit;
/**
 * Atomic ingest of one file: ledger row + records + completion flag, or nothing.
 *
 * ```ts
 * await ingest(client, { runId, filePath, sourceType }, async (unit) => {
 *   if (unit.alreadyIngested) return;
 *   await unit.write(records);
 * });
 * ```
 *
 * Guarantees, matching db_writer.py exactly:
 *
 * - On clean return the ledger row is marked complete with its record count and
 *   the whole unit commits together.
 * - On a thrown error everything rolls back, ledger row included, so the file
 *   has no trace in the ledger and the next run retries it.
 * - `alreadyIngested` is true only when a previous run marked the file complete
 *   — never merely because a row exists.
 * - An abandoned unit (row present, not complete) is reclaimed: orphaned records
 *   are deleted and it is re-ingested under the current run, so a retry cannot
 *   double-count a partial write.
 */
async function ingest(client, params, body) {
    const { runId, filePath, sourceType, rawPayload } = params;
    const fileHash = await computeFileHash(filePath);
    // Fast path for a re-run: the file is already complete, so there is nothing
    // to lock, write, commit or roll back. This read is intentionally outside a
    // transaction, which is safe because ingest_complete is monotonic — it goes
    // false -> true exactly once and never back, so a true observed here cannot
    // be invalidated by a concurrent transaction. A false may be stale, which is
    // why it falls through to the locking upsert rather than being acted on.
    const preflight = await client.query('SELECT ingest_complete FROM ingested_files WHERE file_hash = $1', [fileHash]);
    if (preflight.rows[0]?.ingest_complete) {
        // No BEGIN was issued, and node-postgres autocommits, so unlike the Python
        // version there is no implicit transaction left open by that SELECT and
        // nothing to release here.
        const unit = new IngestUnit(client, runId, filePath, fileHash, true);
        return { fileHash, alreadyIngested: true, recordsWritten: 0, value: await body(unit) };
    }
    await client.query('BEGIN');
    try {
        // Upsert-and-lock. Returns the row's completion state either way. A row
        // that is already complete keeps every original value: a finished ingest is
        // immutable audit data and must not be re-stamped with a later run's id.
        const claimed = await client.query(`INSERT INTO ingested_files
         (file_hash, run_id, file_path, file_name, source_type, raw_payload)
       VALUES ($1, $2, $3, $4, $5, $6)
       ON CONFLICT (file_hash) DO UPDATE SET
         run_id      = CASE WHEN ingested_files.ingest_complete
                            THEN ingested_files.run_id      ELSE EXCLUDED.run_id      END,
         file_path   = CASE WHEN ingested_files.ingest_complete
                            THEN ingested_files.file_path   ELSE EXCLUDED.file_path   END,
         file_name   = CASE WHEN ingested_files.ingest_complete
                            THEN ingested_files.file_name   ELSE EXCLUDED.file_name   END,
         source_type = CASE WHEN ingested_files.ingest_complete
                            THEN ingested_files.source_type ELSE EXCLUDED.source_type END,
         raw_payload = CASE WHEN ingested_files.ingest_complete
                            THEN ingested_files.raw_payload ELSE EXCLUDED.raw_payload END,
         ingested_at = CASE WHEN ingested_files.ingest_complete
                            THEN ingested_files.ingested_at ELSE now()                END
       RETURNING ingest_complete`, [
            fileHash,
            runId,
            filePath,
            filePath.split(/[\\/]/).pop() ?? filePath,
            sourceType,
            rawPayload ?? {},
        ]);
        const alreadyIngested = claimed.rows[0]?.ingest_complete === true;
        if (!alreadyIngested) {
            // Reclaim an abandoned unit. A no-op for a row just inserted; for a
            // stranded one it clears partial records so this run's write is the only
            // contribution.
            await client.query('DELETE FROM forensic_records WHERE file_hash = $1', [fileHash]);
        }
        const unit = new IngestUnit(client, runId, filePath, fileHash, alreadyIngested);
        const value = await body(unit);
        if (alreadyIngested) {
            // Another process completed this file between the preflight read and the
            // upsert. Nothing was written and nothing should be; release the row lock.
            await client.query('ROLLBACK');
            return { fileHash, alreadyIngested: true, recordsWritten: 0, value };
        }
        await client.query(`UPDATE ingested_files
          SET ingest_complete = TRUE,
              record_count    = $1,
              completed_at    = now()
        WHERE file_hash = $2`, [unit.recordsWritten, fileHash]);
        await client.query('COMMIT');
        return { fileHash, alreadyIngested: false, recordsWritten: unit.recordsWritten, value };
    }
    catch (error) {
        try {
            await client.query('ROLLBACK');
        }
        catch (rollbackError) {
            // Surface the original failure, not the rollback's. A rollback that fails
            // usually means the connection is already gone, in which case the server
            // aborts the transaction anyway and the atomicity guarantee still holds.
            console.error('[dbWriter] ROLLBACK failed after an ingest error:', rollbackError);
        }
        throw error;
    }
}
/**
 * Insert one validated NormalizedRecord into forensic_records.
 *
 * Does NOT manage a transaction — the caller owns the boundary, normally by
 * being inside an `ingest()` unit.
 */
async function writeRecord(client, runId, fileHash, record) {
    const validated = contracts_1.NormalizedRecord.parse(record);
    await client.query(`INSERT INTO forensic_records
      (file_hash, run_id, incident_id, source_type, event_time,
       bug_type, process_name, pid, bundle_id, fields)
     VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)`, [
        fileHash,
        runId,
        validated.incident_id ?? null,
        validated.source_type,
        validated.event_time ?? null,
        validated.bug_type ?? null,
        validated.process_name ?? null,
        validated.pid ?? null,
        validated.bundle_id ?? null,
        validated.fields,
    ]);
}
/**
 * Bulk form of writeRecord. Does NOT manage a transaction.
 *
 * Chunked, because the previous version built a single statement with ten
 * placeholders per record. Postgres caps bind parameters at 65535, so any batch
 * over 6553 records failed outright — reachable with one ordinary gcloud log
 * export, and the kind of failure that, before `ingest()`, would have left a
 * committed ledger row and no records.
 */
const MAX_PARAMS_PER_STATEMENT = 65535;
const COLUMNS_PER_RECORD = 10;
exports.MAX_RECORDS_PER_STATEMENT = Math.floor(MAX_PARAMS_PER_STATEMENT / COLUMNS_PER_RECORD);
async function writeRecords(client, runId, fileHash, records) {
    if (records.length === 0) {
        return 0;
    }
    for (let offset = 0; offset < records.length; offset += exports.MAX_RECORDS_PER_STATEMENT) {
        const chunk = records.slice(offset, offset + exports.MAX_RECORDS_PER_STATEMENT);
        const values = [];
        const params = [];
        chunk.forEach((record, index) => {
            const validated = contracts_1.NormalizedRecord.parse(record);
            const base = index * COLUMNS_PER_RECORD;
            values.push(`($${base + 1}, $${base + 2}, $${base + 3}, $${base + 4}, $${base + 5}, $${base + 6}, $${base + 7}, $${base + 8}, $${base + 9}, $${base + 10})`);
            params.push(fileHash, runId, validated.incident_id ?? null, validated.source_type, validated.event_time ?? null, validated.bug_type ?? null, validated.process_name ?? null, validated.pid ?? null, validated.bundle_id ?? null, validated.fields);
        });
        await client.query(`INSERT INTO forensic_records
        (file_hash, run_id, incident_id, source_type, event_time,
         bug_type, process_name, pid, bundle_id, fields)
       VALUES ${values.join(', ')}`, params);
    }
    return records.length;
}
/**
 * Ledger rows that were started and never finished.
 *
 * Expected to be empty. Non-empty means a hard kill mid-unit, or rows the 0002
 * migration could not classify — files predating the atomicity fix that have no
 * records, which may be legitimately empty or may be ones the old code lost.
 * Either way the next run retries them; this exists so an operator can see them
 * rather than infer them, since the failure mode being fixed never announced
 * itself.
 */
async function incompleteIngests(client) {
    const result = await client.query(`SELECT file_hash, file_path
       FROM ingested_files
      WHERE NOT ingest_complete
      ORDER BY ingested_at`);
    return result.rows.map((row) => ({ fileHash: row.file_hash, filePath: row.file_path }));
}
